### PR TITLE
feat: config option to make exchange rate polling more forgiving

### LINKED ISSUE #304: Tolerate failed rate backend requests
Right now, rates are updated at the same interval they're invalidated as too old (60 seconds, by default). The problem is any single failed request results in the connector being unavailable until the next request succeeds, since existing rates are discarded if the request fails. Rates should be updated more frequently than their maximum acceptable age to tolerate one-off request failures with the rate backend.

_Originally posted by @kincaidoneil in https://github.com/interledger-rs/interledger-rs/pull/295_

### PR DESCRIPTION
**Pull request by bstrie**
closes #304

### CODE REVIEW

File: crates/ilp-node/src/node.rs [hunk 1]
```diff
@@ -183,6 +186,10 @@ pub struct InterledgerNode {
     /// Defaults to 60000ms (60 seconds).
     #[serde(default = "default_exchange_rate_poll_interval")]
     pub exchange_rate_poll_interval: u64,
+    /// The number of consecutive failed polls to the exchange rate provider
+    /// that the connector will tolerate before invalidating the exchange rate cache.
+    #[serde(default = "default_exchange_rate_poll_lenience")]
+    pub exchange_rate_poll_lenience: u32,
```
Comment (emschwartz) [inline crates/ilp-node/src/node.rs]: How about `exchange_rate_poll_failure_tolerance`?

File: crates/interledger-api/tests/node_settings.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ fn node_settings_test() {
         secret_seed: random_secret(),
         route_broadcast_interval: Some(200),
         exchange_rate_poll_interval: 60000,
+        exchange_rate_failed_polls_before_invalidation: 5,
```
Comment (emschwartz) [inline crates/interledger-api/tests/node_settings.rs]: The names of the variable in the tests need to be updated

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 3]
```diff
@@ -8,12 +8,14 @@ use interledger_service::*;
 // TODO remove the dependency on interledger_settlement, that doesn't really make sense for this minor import
 use interledger_settlement::{Convert, ConvertDetails};
 use log::{debug, error, trace, warn};
+use parking_lot::RwLock;
```
Comment (emschwartz) [inline crates/interledger-service-util/src/exchange_rates_service.rs]: Can we just use an AtomicU32 instead?

Comment (bstrie) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: Sure, though that wouldn't solve the potential race condition.

Comment (emschwartz) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: I'm not too worried about that particular race condition

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 4]
```diff
@@ -270,10 +272,10 @@ where
             .map_err(move |_| {
                 // Note that a race between the read on this line and the check on the line after
                 // is quite unlikely as long as the interval between polls is reasonable.
-                let failed_polls: u32 = { *consecutive_failed_polls.read() };
+                let failed_polls = consecutive_failed_polls.load(Ordering::Relaxed);
```
Comment (emschwartz) [inline crates/interledger-service-util/src/exchange_rates_service.rs]: What about just using [`fetch_add`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.fetch_add) and then checking the return. Whether or not the `failed_polls < failed_polls_before_invalidation`, the number of failed polls has gone up

**Pull request merged**